virtual/server.go

```diff
@@ -348,3 +335,16 @@ func copyResultIntoStreamAndCloseResult(
 		terminateConnection(w)
 	}
 }
+
+func getContextFromRequest(r *http.Request) context.Context {
```

Comment:
If there is no deadline on the context, can you add a default one?  I think 15s would be good.